pkg/util/kubernetes/apiserver/hpa_controler_test.go

```diff
@@ -0,0 +1,302 @@
+	informerFactory := informers.NewSharedInformerFactory(client, 0)
+
+	autoscalerController := NewAutoscalerController(
+		client,
+		itf,
+		dcl,
+		informerFactory.Autoscaling().V2beta1().HorizontalPodAutoscalers(),
+	)
+	autoscalerController.autoscalersListerSynced = alwaysReady
+
+	return autoscalerController, informerFactory
+}
+
+type leItf struct{}
+
+func (le *leItf) IsLeader() bool { return true }
+
+type dClItf struct {
+	queryMetricsFunc func(from, to int64, query string) ([]datadog.Series, error)
+}
+
+func (d *dClItf) QueryMetrics(from, to int64, query string) ([]datadog.Series, error) {
+	if d.queryMetricsFunc != nil {
+		return d.queryMetricsFunc(from, to, query)
+	}
+	return nil, nil
+}
+
+// TestAutoscalerController is an integration test of the AutoscalerController
+func TestAutoscalerController(t *testing.T) {
+	client := fake.NewSimpleClientset()
+	i := &leItf{}
+	metricName := "foo"
+	name := custommetrics.GetConfigmapName()
+	ddSeries := []datadog.Series{
+		{
+			Metric: &metricName,
+			Points: []datadog.DataPoint{
+				{1531492452, 12},
+				{1531492486, 14},
+			},
+		},
+	}
+	d := &dClItf{
+		queryMetricsFunc: func(from, to int64, query string) ([]datadog.Series, error) {
+			return ddSeries, nil
+		},
+	}
+	hctrl, inf := newFakeAutoscalerController(client, LeaderElectorItf(i), hpa.DatadogClient(d))
+	hctrl.poller.batchWindow = 600 // Do not trigger the refresh or batch call to avoid flakiness.
+	hctrl.poller.refreshPeriod = 600
+	hctrl.poller.gcPeriodSeconds = 600
+	hctrl.autoscalers = make(chan interface{}, 1)
+
+	stop := make(chan struct{})
+	defer close(stop)
+	inf.Start(stop)
+	go hctrl.Run(stop)
+
+	c := client.AutoscalingV2beta1()
+	require.NotNil(t, c)
+
+	mockedHPA := newFakeHorizontalPodAutoscaler(
+		"hpa_1",
+		"default",
+		"1",
+		"foo",
+		map[string]string{"foo": "bar"},
+	)
+
+	_, err := c.HorizontalPodAutoscalers("default").Create(mockedHPA)
+	require.NoError(t, err)
+	timeout := time.NewTimer(5 * time.Second)
+	ticker := time.NewTicker(500 * time.Millisecond)
+	select {
+	case <-hctrl.autoscalers:
+	case <-timeout.C:
+		require.FailNow(t, "Timeout waiting for HPAs to update")
+	}
+	// Check local cache store is 1:1 with expectations
+	storedHPA, err := hctrl.autoscalersLister.HorizontalPodAutoscalers(mockedHPA.Namespace).Get(mockedHPA.Name)
+	require.NoError(t, err)
+	require.Equal(t, storedHPA, mockedHPA)
+
+	select {
+	case <-ticker.C:
+		hctrl.toStore.m.Lock()
+		st := hctrl.toStore.data
+		hctrl.toStore.m.Unlock()
+		require.NotEmpty(t, st)
+	case <-timeout.C:
+		require.FailNow(t, "Timeout waiting for HPAs to update")
+	}
+	// Forcing update the to global store and clean local cache store
+	hctrl.pushToGlobalStore(hctrl.toStore.data)
+	require.Nil(t, hctrl.toStore.data)
+
+	// Test that the Global store contains the correct data
+	select {
+	case <-ticker.C:
+		k, _ := client.CoreV1().ConfigMaps("default").Get(name, metav1.GetOptions{})
```

Comment:
We should trust that the `Store` is properly unit tested by the `custommetrics` package and only communicate through the `Store` interface. The fact that the `Store` uses a configmap is an implementation detail.

If the `Store` interface is too clunky, consider using a `mockStore` or modifying the `Store` interface.